### PR TITLE
Seed

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
Setting a seed in MM is not trivial and often we need reproducible reults

**Description of the Change:**
The `settings` object now supports the `SEED` attribute, which is random unless it is set by the user. To unset it, just set it equal to `None`.

**Benefits:**
Easy to get reproducible results without messing with numpy.random.

**Possible Drawbacks:**
None?

**Related GitHub Issues:**

### CODE REVIEW

**Review by sduquemesa — COMMENTED** [review]
Thanks @ziofil! This was a long overdue chore. I have few comments:

Is there any specific reason why the MT19937 random number generator is used? This in particular leads to a convoluted way of instantiating and referencing the random number generator and is marked as a legacy method in numpy's [documentation](https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState). Numpy has a simple API [`numpy.random.default_rng`](https://numpy.org/doc/stable/reference/random/generator.html#random-generator) which is also automatically seeded when no seed is provided.

File: mrmustard/__init__.py [hunk 1]
```diff
@@ -46,6 +50,22 @@ def __init__(self):
         self.HOMODYNE_SQUEEZING = 10.0
         # misc
         self.PROGRESSBAR = True
+        self._seed = np.random.randint(0, 2**32)
+        self.random_state = RandomState(MT19937(SeedSequence(self._seed)))
```
Comment (sduquemesa) [inline mrmustard/__init__.py:54]: For the sake of clarity, I believe this should be name `rng` or `random_number_generator` as it provides random numbers and not the state of the generator.

**Comment by ziofil** [comment]
> Thanks @ziofil! This was a long overdue chore. I have few comments:
> 
> Is there any specific reason why the MT19937 random number generator is used? This in particular leads to a convoluted way of instantiating and referencing the random number generator and is marked as a legacy method in numpy's [documentation](https://numpy.org/doc/stable/reference/random/legacy.html#numpy.random.RandomState). Numpy has a simple API [`numpy.random.default_rng`](https://numpy.org/doc/stable/reference/random/generator.html#random-generator) which is also automatically seeded when no seed is provided.

Oh sure, I think I had found that line somewhere in numpy's documentation, but probably it was outdated. Let's change it to whatever is recommended.

File: tests/test_training/test_opt.py [hunk 2]
```diff
@@ -39,16 +39,16 @@
...
 
 
 @given(n=st.integers(0, 3))
 def test_S2gate_coincidence_prob(n):
     """Testing the optimal probability of obtaining |n,n> from a two mode squeezed vacuum"""
-    tf.random.set_seed(137)
     S = S2gate(
-        r=abs(np.random.normal()),
-        phi=np.random.normal(),
+        r=abs(settings.rng.normal()),
```
Comment (sduquemesa) [inline tests/test_training/test_opt.py:50]: Should this test set the seed in the beggining like the other tests as well?

File: tests/test_training/test_opt.py [hunk 3]
```diff
@@ -70,14 +70,13 @@ def test_hong_ou_mandel_optimizer(i, k):
...
     """
-    tf.random.set_seed(137)
     r = np.arcsinh(1.0)
     s2_0, s2_1, bs = (
         S2gate(r=r, phi=0.0, phi_trainable=True)[0, 1],
         S2gate(r=r, phi=0.0, phi_trainable=True)[2, 3],
         BSgate(
-            theta=np.arccos(np.sqrt(k / (i + k))) + 0.1 * np.random.normal(),
-            phi=np.random.normal(),
+            theta=np.arccos(np.sqrt(k / (i + k))) + 0.1 * settings.rng.normal(),
```
Comment (sduquemesa) [inline tests/test_training/test_opt.py:78]: Same comment — should the seed be set in this test as well?

**Review by sduquemesa — CHANGES_REQUESTED** [review]
Thanks @ziofil! I have left some more comments.

File: tests/test_lab/test_states.py [hunk 4]
```diff
@@ -201,7 +201,7 @@ def test_modes_after_projection(m):
 def test_modes_after_double_projection(n, m):
     """Test number of modes is correct after double projection."""
     assume(n != m)
-    a = Gaussian(4) << Fock([1, 2])[n, m]
+    a = Gaussian(4) >> Dgate(x=1.0)[0, 1, 2, 3] << Fock([1, 2])[n, m]
```
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_lab/test_states.py:204]: to make the test a bit more general

File: tests/test_lab/test_states.py [hunk 5]
```diff
@@ -268,3 +268,43 @@ def test_ket_from_pure_dm(n, cutoffs):
 
     # check test state calculated the same ket as the original state
     assert np.allclose(test_ket, fock_state.ket())
+
+
+def test_ket_probability():
```
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_lab/test_states.py:273]: I looked at what lines codecov was complaining about and I added tests to cover them

File: tests/test_physics/test_fock/test_fock.py [hunk 6]
```diff
@@ -189,30 +189,6 @@ def test_dm_to_ket_error():
         dm_to_ket(state)
 
 
-def test_fock_trace_mode1():
```
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_physics/test_fock/test_fock.py:192]: 🤔 I think it happened when I resolved a conflict with the `develop` branch. They shouldn't have been removed.

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -22,6 +22,9 @@
   The result is added to the `callback_history` attribute of the optimizer.
   [(#175)](https://github.com/XanaduAI/MrMustard/pull/175)
 
+  * MrMustard's settings object (accessible via `from mrmustard import settings`) now supports `SEED`. This will give reproducible results whenever randomness is involved. The seed is unset by default, and it can be unset again with `settings.SEED = None`.
+  [(#183)](https://github.com/XanaduAI/MrMustard/pull/183)
```
[the remark this answers is not in the document]
Comment (ziofil) [reply .github/CHANGELOG.md:26]: You mean `settings.rng`? I meant it to be used internally. The only thing a user should worry about is the SEED setting. But ok, I can write a snippet to clarify.

File: tests/test_training/test_opt.py (hunk 2 above)
Comment (ziofil) [reply tests/test_training/test_opt.py:49]: there is a global seed set at the top of the file, but I moved the seeds into each test

**Review by sduquemesa — APPROVED** [review]

**Pull request merged**